In [1]:
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)

In [2]:
credit_data = pd.read_csv('datasets/credit_train.csv')
credit_data.head()

Loan ID                           Customer ID  \
0  14dd8831-6af5-400b-83ec-68e61888a048  981165ec-3274-42f5-a3b4-d104041a9ca9   
1  4771cc26-131a-45db-b5aa-537ea4ba5342  2de017a3-2e01-49cb-a581-08169e83be29   
2  4eed4e6a-aa2f-4c91-8651-ce984ee8fb26  5efb2b2b-bf11-4dfd-a572-3761a2694725   
3  77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a  e777faab-98ae-45af-9a86-7ce5b33b1011   
4  d4062e70-befa-4995-8643-a0de73938182  81536ad9-5ccf-4eb8-befb-47a4d608658e   

  Loan Status  Current Loan Amount        Term  Credit Score  Annual Income  \
0  Fully Paid             445412.0  Short Term         709.0      1167493.0   
1  Fully Paid             262328.0  Short Term           NaN            NaN   
2  Fully Paid           99999999.0  Short Term         741.0      2231892.0   
3  Fully Paid             347666.0   Long Term         721.0       806949.0   
4  Fully Paid             176220.0  Short Term           NaN            NaN   

  Years in current job Home Ownership             Purpose  Monthly Debt  \
0              8 years  Home Mortgage   Home Improvements       5214.74   
1            10+ years  Home Mortgage  Debt Consolidation      33295.98   
2              8 years       Own Home  Debt Consolidation      29200.53   
3              3 years       Own Home  Debt Consolidation       8741.90   
4              5 years           Rent  Debt Consolidation      20639.70   

   Years of Credit History  Months since last delinquent  \
0                     17.2                           NaN   
1                     21.1                           8.0   
2                     14.9                          29.0   
3                     12.0                           NaN   
4                      6.1                           NaN   

   Number of Open Accounts  Number of Credit Problems  Current Credit Balance  \
0                      6.0                        1.0                228190.0   
1                     35.0                        0.0                229976.0   
2                     18.0                        1.0                297996.0   
3                      9.0                        0.0                256329.0   
4                     15.0                        0.0                253460.0   

   Maximum Open Credit  Bankruptcies  Tax Liens  
0             416746.0           1.0        0.0  
1             850784.0           0.0        0.0  
2             750090.0           0.0        0.0  
3             386958.0           0.0        0.0  
4             427174.0           0.0        0.0

In [3]:
credit_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100514 entries, 0 to 100513
Data columns (total 19 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Loan ID                       100000 non-null  object 
 1   Customer ID                   100000 non-null  object 
 2   Loan Status                   100000 non-null  object 
 3   Current Loan Amount           100000 non-null  float64
 4   Term                          100000 non-null  object 
 5   Credit Score                  80846 non-null   float64
 6   Annual Income                 80846 non-null   float64
 7   Years in current job          95778 non-null   object 
 8   Home Ownership                100000 non-null  object 
 9   Purpose                       100000 non-null  object 
 10  Monthly Debt                  100000 non-null  float64
 11  Years of Credit History       100000 non-null  float64
 12  Months since last delinquent  46859 non-null

In [4]:
credit_data.describe(include="all")

Loan ID  \
count                                 100000   
unique                                 81999   
top     14dd8831-6af5-400b-83ec-68e61888a048   
freq                                       2   
mean                                     NaN   
std                                      NaN   
min                                      NaN   
25%                                      NaN   
50%                                      NaN   
75%                                      NaN   
max                                      NaN   

                                 Customer ID Loan Status  Current Loan Amount  \
count                                 100000      100000         1.000000e+05   
unique                                 81999           2                  NaN   
top     981165ec-3274-42f5-a3b4-d104041a9ca9  Fully Paid                  NaN   
freq                                       2       77361                  NaN   
mean                                     NaN         NaN         1.176045e+07   
std                                      NaN         NaN         3.178394e+07   
min                                      NaN         NaN         1.080200e+04   
25%                                      NaN         NaN         1.796520e+05   
50%                                      NaN         NaN         3.122460e+05   
75%                                      NaN         NaN         5.249420e+05   
max                                      NaN         NaN         1.000000e+08   

              Term  Credit Score  Annual Income Years in current job  \
count       100000  80846.000000   8.084600e+04                95778   
unique           2           NaN            NaN                   11   
top     Short Term           NaN            NaN            10+ years   
freq         72208           NaN            NaN                31121   
mean           NaN   1076.456089   1.378277e+06                  NaN   
std            NaN   1475.403791   1.081360e+06                  NaN   
min            NaN    585.000000   7.662700e+04                  NaN   
25%            NaN    705.000000   8.488440e+05                  NaN   
50%            NaN    724.000000   1.174162e+06                  NaN   
75%            NaN    741.000000   1.650663e+06                  NaN   
max            NaN   7510.000000   1.655574e+08                  NaN   

       Home Ownership             Purpose   Monthly Debt  \
count          100000              100000  100000.000000   
unique              4                  16            NaN   
top     Home Mortgage  Debt Consolidation            NaN   
freq            48410               78552            NaN   
mean              NaN                 NaN   18472.412336   
std               NaN                 NaN   12174.992609   
min               NaN                 NaN       0.000000   
25%               NaN                 NaN   10214.162500   
50%               NaN                 NaN   16220.300000   
75%               NaN                 NaN   24012.057500   
max               NaN                 NaN  435843.280000   

        Years of Credit History  Months since last delinquent  \
count             100000.000000                  46859.000000   
unique                      NaN                           NaN   
top                         NaN                           NaN   
freq                        NaN                           NaN   
mean                  18.199141                     34.901321   
std                    7.015324                     21.997829   
min                    3.600000                      0.000000   
25%                   13.500000                     16.000000   
50%                   16.900000                     32.000000   
75%                   21.700000                     51.000000   
max                   70.500000                    176.000000   

        Number of Open Accounts  Number of Credit Problems  \
count              100000.00000              100000.000000

### Removing columns that are not important for our analysis

In [5]:
credit_data = credit_data.drop(columns=['Loan ID', 'Customer ID', 'Months since last delinquent'])
credit_data.columns

Index(['Loan Status', 'Current Loan Amount', 'Term', 'Credit Score',
       'Annual Income', 'Years in current job', 'Home Ownership', 'Purpose',
       'Monthly Debt', 'Years of Credit History', 'Number of Open Accounts',
       'Number of Credit Problems', 'Current Credit Balance',
       'Maximum Open Credit', 'Bankruptcies', 'Tax Liens'],
      dtype='object')

We removed the columns 'Loan ID' and 'Customer ID' as they do not provide any insights; they are just random unique values. We dropped 'Months since last delinquent' since it contains a large proportion of NA values and it is not very relevant to our topic of analysis.

### Removing rows that have missing data

In [6]:
credit_data = credit_data.dropna()
credit_data.head()

Loan Status  Current Loan Amount        Term  Credit Score  Annual Income  \
0   Fully Paid             445412.0  Short Term         709.0      1167493.0   
2   Fully Paid           99999999.0  Short Term         741.0      2231892.0   
3   Fully Paid             347666.0   Long Term         721.0       806949.0   
5  Charged Off             206602.0  Short Term        7290.0       896857.0   
6   Fully Paid             217646.0  Short Term         730.0      1184194.0   

  Years in current job Home Ownership             Purpose  Monthly Debt  \
0              8 years  Home Mortgage   Home Improvements       5214.74   
2              8 years       Own Home  Debt Consolidation      29200.53   
3              3 years       Own Home  Debt Consolidation       8741.90   
5            10+ years  Home Mortgage  Debt Consolidation      16367.74   
6             < 1 year  Home Mortgage  Debt Consolidation      10855.08   

   Years of Credit History  Number of Open Accounts  \
0                     17.2                      6.0   
2                     14.9                     18.0   
3                     12.0                      9.0   
5                     17.3                      6.0   
6                     19.6                     13.0   

   Number of Credit Problems  Current Credit Balance  Maximum Open Credit  \
0                        1.0                228190.0             416746.0   
2                        1.0                297996.0             750090.0   
3                        0.0                256329.0             386958.0   
5                        0.0                215308.0             272448.0   
6                        1.0                122170.0             272052.0   

   Bankruptcies  Tax Liens  
0           1.0        0.0  
2           0.0        0.0  
3           0.0        0.0  
5           0.0        0.0  
6           1.0        0.0

In [7]:
credit_data.shape

(77271, 16)

### Checking for and removing duplicates

In [8]:
credit_data.duplicated()

0        False
2        False
3        False
5        False
6        False
         ...  
99994    False
99995     True
99996    False
99997    False
99998    False
Length: 77271, dtype: bool

In [9]:
credit_data = credit_data.drop_duplicates()
credit_data.shape

(67490, 16)

### Removing outliers/anomalous data

In [10]:
credit_data = credit_data[credit_data['Current Loan Amount']!=99999999]
credit_data.shape

(56461, 16)

We noticed there is a placeholder of 99999999 for unknown 'Current Loan Amount', so we removed it from the dataset.

In [11]:
credit_data = credit_data[credit_data['Credit Score']<=850]
credit_data = credit_data[credit_data['Credit Score']>=300]
credit_data.shape

(52167, 16)

We noticed there were outlier values for 'Credit Score'. The base FICO® Scores range from 300 to 850, so we removed anything out of this range.

In [12]:
Q3 = credit_data['Maximum Open Credit'].quantile(.75)
Q1 = credit_data['Maximum Open Credit'].quantile(.25)
credit_data = credit_data[credit_data['Maximum Open Credit']<=(Q3+1.5*(Q3-Q1))]
credit_data['Maximum Open Credit'].skew()

0.8766170181905047

We noticed there were high outliers for 'Maximum Open Credit', with the maximum value (1.539738e+09) being much higher than the 3rd quartile (7.829580e+05). We used the Interquartile Range method (1.5 * IQR) to remove the higher bound outliers. Skewness of 0.8766170181905047 is moderately skewed and acceptable.

### Standardising names

In [13]:
credit_data['Purpose'].replace('other', 'Other', inplace=True)
credit_data['Purpose'].replace('small_business', 'Small Business', inplace=True)
credit_data['Purpose'].replace('major_purchase', 'Major Purchase', inplace=True)
credit_data['Purpose'].replace('renewable_energy', 'Renewable Energy', inplace=True)
credit_data['Purpose'].replace('moving', 'Moving', inplace=True)
credit_data['Purpose'].replace('vacation', 'Vacation', inplace=True)
credit_data['Purpose'].replace('wedding', 'Wedding', inplace=True)
credit_data['Home Ownership'].replace('HaveMortgage', 'Have Mortgage', inplace=True)

We noticed in 'Purpose' the category names 'other' and 'Other' were not standardised, so we standardised it to 'Other'. We also standardised the other category names in 'Purpose' and 'Home Ownership for better organisation.

In [14]:
credit_data.to_csv('datasets/cleaned_credit.csv', index=False)